python/ray/rllib/models/lstm.py

```diff
@@ -2,56 +2,173 @@
 from __future__ import division
 from __future__ import print_function
 
+"""LSTM support for RLlib.
+
+The main trick here is that we add the time dimension at the last moment.
+The non-LSTM layers of the model see their inputs as one flat batch. Before
+the LSTM cell, we reshape the input to add the expected time dimension. During
+postprocessing, we dynamically pad the experience batches so that this
+reshaping is possible.
+
+See the add_time_dimension() and chop_into_sequences() functions below for
+more info.
+"""
+
+
 import numpy as np
 import tensorflow as tf
 import tensorflow.contrib.rnn as rnn
 import distutils.version
 
-from ray.rllib.models.misc import (conv2d, linear, flatten,
-                                   normc_initializer)
+from ray.rllib.models.misc import linear, normc_initializer
 from ray.rllib.models.model import Model
 
 
+def add_time_dimension(padded_inputs, seq_lens):
+    """Adds a time dimension to padded inputs.
+
+    Arguments:
+        padded_inputs (Tensor): a padded batch of sequences. That is,
+            for seq_lens=[1, 2, 2], then inputs=[A, *, B, B, C, C], where
```

Comment:
When will you have situations where you pad in the middle of the given sequence? ie, only during batch SGD where your batch has N trajectories?

Reply:
Yes, in general you will have to pad when-ever you have variable length trajectories and they aren't all divisible by max_seq_len